contracts/router/LinkedPool.sol

```diff
@@ -0,0 +1,383 @@
+    /// @param nodeIndex        The index of the node to which the pool will be added
+    /// @param pool             The address of the pool
+    /// @param poolModule       The address of the pool module
+    function addPool(
+        uint256 nodeIndex,
+        address pool,
+        address poolModule
+    ) external onlyOwner {
+        require(pool != address(0), "Pool address can't be zero");
+        _addPool(nodeIndex, pool, poolModule);
+    }
+
+    /// @inheritdoc ILinkedPool
+    function swap(
+        uint8 nodeIndexFrom,
+        uint8 nodeIndexTo,
+        uint256 dx,
+        uint256 minDy,
+        uint256 deadline
+    ) external returns (uint256 amountOut) {
+        uint256 totalTokens = _nodes.length;
+        // solhint-disable-next-line not-rely-on-time
+        require(block.timestamp <= deadline, "Deadline not met");
+        require(
+            nodeIndexFrom < totalTokens && nodeIndexTo < totalTokens && nodeIndexFrom != nodeIndexTo,
+            "Swap not supported"
+        );
+        // Pull initial token from the user. LinkedPool assumes that the tokens have no transfer fees enabled,
+        // thus the balance checks are omitted.
+        address tokenIn = _nodes[nodeIndexFrom].token;
+        IERC20(tokenIn).safeTransferFrom(msg.sender, address(this), dx);
+        amountOut = _multiSwap(nodeIndexFrom, nodeIndexTo, dx).amountOut;
+        require(amountOut >= minDy, "Swap didn't result in min tokens");
+        // Transfer the tokens to the user
+        IERC20(_nodes[nodeIndexTo].token).safeTransfer(msg.sender, amountOut);
+    }
+
+    // ═══════════════════════════════════════════════════ VIEWS ═══════════════════════════════════════════════════════
+
+    /// Note: this calculates a quote for a predefined swap path between two tokens. If any of the tokens is
+    /// presented more than once in the internal tree, there might be a better quote. Integration should use
+    /// findBestPath() instead. This function is present for backwards compatibility.
+    /// @inheritdoc ILinkedPool
+    function calculateSwap(
+        uint8 nodeIndexFrom,
+        uint8 nodeIndexTo,
+        uint256 dx
+    ) external view returns (uint256 amountOut) {
+        uint256 totalTokens = _nodes.length;
+        // Check that the token indexes are within range
+        if (nodeIndexFrom >= totalTokens || nodeIndexTo >= totalTokens) {
+            return 0;
+        }
+        // Check that the token indexes are not the same
+        if (nodeIndexFrom == nodeIndexTo) {
+            return 0;
+        }
+        // Calculate the quote by following the path from "tokenFrom" node to "tokenTo" node in the stored tree
+        // This function might be called by Synapse:Bridge before the swap, so we don't waste gas checking if pool is paused,
+        // as the swap will fail anyway if it is.
+        amountOut = _getMultiSwapQuote({
```

Comment:
tests of `_getMultiSwapQuote` via `calculateSwap` calls?

Not finding them in `LinkedPool.t.sol` except for edge case test of `test_calculateSwap_samePoolTwice`

Reply:
Wouldn't hurt having them, will add in the next PRs

Reply:
@mikeyrf Added tests in 9fcc1c1c679d6cb1cc1221f73c0400b9ebe3b8d6